File: projects/group/legacy-flutter/app/lib/agents/bank_account.dart
import 'package:meta/meta.dart';
import 'package:riverpod/riverpod.dart';

import '../scenarios/events.dart';

enum TxnCategory { transfer, bill, salary, shopping, other }

@immutable
class BankTransaction {
  const BankTransaction({
    required this.id,
    required this.timestamp,
    required this.label,
    required this.amountHkd,
    required this.category,
    this.account,
    this.newPayee = false,
  });

  final String id;
  final DateTime timestamp;
  final String label;
  final String? account;
  final double amountHkd; // negative = debit, positive = credit
  final TxnCategory category;
  final bool newPayee;
}

@immutable
class BankAccountState {
  const BankAccountState({
    required this.balanceHkd,
    required this.history,
  });

  final double balanceHkd;
  final List<BankTransaction> history; // newest first
}

/// In-memory bank account state. The demo starts with a seeded history
/// so the Bank screen looks lived-in. Transfer/payBill calls mutate
/// balance + prepend a new history entry.
class BankAccount extends Notifier<BankAccountState> {
  static const _startingBalance = 482530.55;

  @override
  BankAccountState build() {
    final seed = DateTime.now();
    return BankAccountState(
      balanceHkd: _startingBalance,
      history: [
        BankTransaction(
          id: 'seed_pension',
          timestamp: seed.subtract(const Duration(days: 2)),
          label: 'Pension — MPF',
          amountHkd: 8500.00,
          category: TxnCategory.salary,
        ),
        BankTransaction(
          id: 'seed_groceries',
          timestamp: seed.subtract(const Duration(days: 3)),
          label: 'Groceries · Wellcome',
          amountHkd: -412.50,
          category: TxnCategory.shopping,
        ),
        BankTransaction(
          id: 'seed_son',
          timestamp: seed.subtract(const Duration(days: 5)),
          label: 'Son (David Wong)',
          amountHkd: -2000.00,
          category: TxnCategory.transfer,
        ),
        BankTransaction(
          id: 'seed_clp',
          timestamp: seed.subtract(const Duration(days: 10)),
          label: 'CLP Power HK Ltd',
          amountHkd: -412.00,
          category: TxnCategory.bill,
        ),
      ],
    );
  }

  /// Commit a confirmed transfer to the ledger. Call this only after
  /// Guardian has decided NOT to block the transaction (no blocking
  /// intervention pending). Balance is decreased; a history row is
  /// prepended so the bank screen reflects the change immediately.
  BankTransaction commitTransfer(TransactionEvent event) {
    final txn = BankTransaction(
      id: 'txn_${event.id}',
      timestamp: event.timestamp,
      label: event.toName,
      account: event.toAccount,
      amountHkd: -event.amountHkd,
      category: TxnCategory.transfer,
      newPayee: event.newRecipient,
    );
    state = BankAccountState(
      balanceHkd: state.balanceHkd - event.amountHkd,
      history: [txn, ...state.history],
    );
    return txn;
  }

  /// Pay a utility / recurring bill. No Guardian scoring (bill payments
  /// to whitelisted utilities are considered safe by default).
  BankTransaction payBill({required String name, required double amount}) {
    final txn = BankTransaction(
      id: 'bill_${DateTime.now().microsecondsSinceEpoch}',
      timestamp: DateTime.now(),
      label: name,
      amountHkd: -amount,
      category: TxnCategory.bill,
    );
    state = BankAccountState(
      balanceHkd: state.balanceHkd - amount,
      history: [txn, ...state.history],
    );
    return txn;
  }

  /// Test / demo helper: reset to the seeded state.
  void reset() {
    state = build();
  }
}

final bankAccountProvider =
    NotifierProvider<BankAccount, BankAccountState>(BankAccount.new);

File: projects/group/streamlit_app/guardian/agents/bank_account.py
"""Mock bank account ledger. 1:1 port of ``app/lib/agents/bank_account.dart``.

Starts with a small seeded transaction history so the Bank screen looks
lived-in. ``commit_transfer`` mutates the balance + prepends a new row;
``pay_bill`` does the same for utility payments.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timedelta
from enum import Enum

from guardian.scenarios.events import TransactionEvent


class TxnCategory(str, Enum):
    TRANSFER = "transfer"
    BILL = "bill"
    SALARY = "salary"
    SHOPPING = "shopping"
    OTHER = "other"


@dataclass(frozen=True)
class BankTransaction:
    id: str
    timestamp: datetime
    label: str
    amount_hkd: float  # negative = debit, positive = credit
    category: TxnCategory
    account: str | None = None
    new_payee: bool = False


_STARTING_BALANCE = 482_530.55


@dataclass
class BankAccountState:
    balance_hkd: float
    history: list[BankTransaction] = field(default_factory=list)  # newest first


class BankAccount:
    def __init__(self) -> None:
        seed = datetime.now()
        self._state = BankAccountState(
            balance_hkd=_STARTING_BALANCE,
            history=[
                BankTransaction(
                    id="seed_pension",
                    timestamp=seed - timedelta(days=2),
                    label="Pension — MPF",
                    amount_hkd=8500.00,
                    category=TxnCategory.SALARY,
                ),
                BankTransaction(
                    id="seed_groceries",
                    timestamp=seed - timedelta(days=3),
                    label="Groceries · Wellcome",
                    amount_hkd=-412.50,
                    category=TxnCategory.SHOPPING,
                ),
                BankTransaction(
                    id="seed_son",
                    timestamp=seed - timedelta(days=5),
                    label="Son (David Wong)",
                    amount_hkd=-2000.00,
                    category=TxnCategory.TRANSFER,
                ),
                BankTransaction(
                    id="seed_clp",
                    timestamp=seed - timedelta(days=10),
                    label="CLP Power HK Ltd",
                    amount_hkd=-412.00,
                    category=TxnCategory.BILL,
                ),
            ],
        )

    @property
    def state(self) -> BankAccountState:
        return self._state

    def commit_transfer(self, event: TransactionEvent) -> BankTransaction:
        txn = BankTransaction(
            id=f"txn_{event.id}",
            timestamp=event.timestamp,
            label=event.to_name,
            account=event.to_account,
            amount_hkd=-event.amount_hkd,
            category=TxnCategory.TRANSFER,
            new_payee=event.new_recipient,
        )
        self._state = BankAccountState(
            balance_hkd=self._state.balance_hkd - event.amount_hkd,
            history=[txn, *self._state.history],
        )
        return txn

    def pay_bill(self, name: str, amount: float) -> BankTransaction:
        txn = BankTransaction(
            id=f"bill_{datetime.now().timestamp():.0f}",
            timestamp=datetime.now(),
            label=name,
            amount_hkd=-amount,
            category=TxnCategory.BILL,
        )
        self._state = BankAccountState(
            balance_hkd=self._state.balance_hkd - amount,
            history=[txn, *self._state.history],
        )
        return txn

    def reset(self) -> None:
        self.__init__()  # re-seed
